src/aws-util/src/s3_uploader.rs

```diff
@@ -0,0 +1,554 @@
+        self.file_size_limit
+    }
+
+    fn part_size_limit(&self) -> u64 {
+        self.part_size_limit
+    }
+}
+
+impl Default for S3MultiPartUploaderConfig {
+    fn default() -> Self {
+        Self {
+            part_size_limit: Self::DEFAULT_PART_SIZE_LIMIT.as_u64(),
+            file_size_limit: Self::DEFAULT_MAX_FILE_SIZE.as_u64(),
+        }
+    }
+}
+
+impl S3MultiPartUploader {
+    /// Creates a an instance of `S3MultiPartUploader` for the given `bucket` and `path`.
+    /// This starts the multi part upload by making a `create_multipart_upload` call, and
+    /// initializes all the internal state required to track the ongoing upload.
+    pub async fn try_new(
+        sdk_config: &SdkConfig,
+        bucket: String,
+        path: String,
+        config: S3MultiPartUploaderConfig,
+    ) -> Result<S3MultiPartUploader, S3MultiPartUploadError> {
+        let client = crate::s3::new_client(sdk_config);
+        let res = client
+            .create_multipart_upload()
+            .bucket(&bucket)
+            .key(&path)
+            .send()
+            .await?;
+        let upload_id = res
+            .upload_id()
+            .ok_or(anyhow!(
+                "create_multipart_upload response missing upload id"
+            ))?
+            .to_string();
+        Ok(S3MultiPartUploader {
+            client,
+            bucket,
+            path,
+            upload_id,
+            part_count: 0,
+            etags: Default::default(),
+            total_bytes_uploaded: 0,
+            buffer: Default::default(),
+            config,
+        })
+    }
+
+    /// Adds the `data` to the internal buffer and flushes the buffer if it is more than
+    /// the part threshold defined in `S3MultiPartUploaderConfig`.
+    /// Note, the size of the data cannot exceed the possible total size of one single file,
+    /// returned by `config.file_size_limit()`.
+    pub async fn add_chunk(
+        &mut self,
+        data: Bytes,
+    ) -> Result<InProgressUpload, S3MultiPartUploadError> {
+        let num_of_bytes: u64 = data.len().try_into().expect("usize to u64");
+        // We should not exceed the user configured `file_size_limit` for this upload.
+        // `remaining_bytes_limit` is defined as `file_size_limit` - `bytes_uploaded` - `bytes_buffered`.
+        if num_of_bytes > self.remaining_bytes_limit() {
+            return Err(S3MultiPartUploadError::ExceedsConfiguredMaxUploadSize {
+                chunk_size: num_of_bytes,
+                remaining_bytes: self.remaining_bytes_limit(),
+                file_size_limit: self.config.file_size_limit(),
+            });
+        }
+        self.buffer.extend_from_slice(&data);
+        self.flush_chunks().await?;
+
+        Ok(InProgressUpload {
+            part_count: self.part_count.try_into().expect("i32 to u32"),
+            bytes_uploaded: self.total_bytes_uploaded,
+            bytes_buffered: self.buffer.len().try_into().expect("usize to u64"),
+        })
```

Comment:
Not sure that returning this `InProgressUpload` is the ideal API for this! I think it'd be slightly nicer if you exposed accessor methods for `part_count`, `bytes_uploaded`, and `bytes_buffered`, and then the caller could access them whenever they liked, rather than having to keep track of the latest `InProgressUpload`?

Reply:
Done, I am exposing the `remaining_bytes()` as a public method instead.